cpp/include/raft/sparse/matrix/detail/diagonal.cuh

```diff
@@ -151,4 +154,126 @@ void set_diagonal(raft::resources const& res,
                    });
 }
 
+/**
+ * @brief Get the diagonal vector from a COO matrix
+ *
+ * This function extracts the diagonal elements from a COO matrix and stores them in a vector.
+ * The diagonal elements are the elements where the row index and column index are the same.
+ *
+ * @tparam T The data type of the matrix elements
+ * @tparam RowType The data type of the row indices
+ * @tparam ColType The data type of the column indices
+ * @tparam NNZType The data type for representing nonzero counts
+ *
+ * @param[in]  res             RAFT resources for managing device memory and streams
+ * @param[in]  coo_matrix_view View of the input COO matrix from which to extract the diagonal
+ * @param[out] diagonal        View of the output vector where diagonal elements will be stored
+ */
+template <typename T, typename RowType, typename ColType, typename NNZType>
+void diagonal(raft::resources const& res,
+              raft::device_coo_matrix_view<T, RowType, ColType, NNZType> coo_matrix_view,
+              raft::device_vector_view<T, RowType> diagonal)
+{
+  auto structure = coo_matrix_view.structure_view();
+  auto nnz       = structure.get_nnz();
+  auto n_rows    = structure.get_n_rows();
+
+  auto values   = coo_matrix_view.get_elements().data();
+  auto rows     = structure.get_rows().data();
+  auto cols     = structure.get_cols().data();
+  auto diag_ptr = diagonal.data_handle();
+
+  raft::matrix::fill(res, diagonal, T(0));
+
+  auto policy = raft::resource::get_thrust_policy(res);
+
+  thrust::for_each(policy,
```

Comment:
Wherever possible, please avoid `thrust::for_each` and use `raft::linalg::unary_op` / `map_offset` or similar.